pingcap/tidb-binlog
# feed downstream by relay log at startup if need

### What problem does this PR solve? 
support to recover at startup by relay log. 
follows #847 #849 
### What is changed and how it works?
- add a status in checkpoint
    - we can learn from the status that there are dirty data written into downstream behind commit ts or not.
- group the config items about relay
    - remove the `relay-read-buf-size`
    - rename `relay-log-size` to `max-file-size`
- make relay.Reader return the origin format  directly
    - this will be more flexible by the user of the reader
- feed downstream by relay log at startup if need.
    - mainly implement in `drainer/relay.go`
    - only fails to init the pd client will try this or just assume the upstream cluster is normal and works like before.

   

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
- kill -9 drainer & pd while insert data into upstream
- check status is Running
- startup drainer
- check recover and status updated to be Normal



Side effects



Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note
